src/test/util/setup_common.cpp

```diff
@@ -252,20 +252,55 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
     }
 
     BlockValidationState state;
-    if (!::ChainstateActive().ActivateBestChain(state, chainparams)) {
+    if (!::ChainstateActive().ActivateBestChain(state)) {
         throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", state.ToString()));
     }
 }
 
-TestChainSetup::TestChainSetup(int blockCount, const std::vector<const char*>& extra_args)
+TestChainSetup::TestChainSetup(int num_blocks, const std::vector<const char*>& extra_args)
     : RegTestingSetup(extra_args)
 {
-    // Generate a 100-block chain:
-    coinbaseKey.MakeNewKey(true);
+    SetMockTime(1598887952);
+    constexpr std::array<unsigned char, 32> vchKey = {
+        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}};
+    coinbaseKey.Set(vchKey.begin(), vchKey.end(), true);
+
+    // Generate a num_blocks length chain:
+    this->mineBlocks(num_blocks);
+
+    CCheckpointData checkpoints{
+        {
+            /* TestChainDATSetup */
+            {   98, uint256S("0x150e127929d578d8129b77a6cb7e2e343a1379aa3feaaa9cce59e0a645756a81") },
+            /* TestChain100Setup */
+            {  100, uint256S("0x6ffb83129c19ebdf1ae3771be6a67fe34b35f4c956326b9ba152fac1649f65ae") },
+            /* TestChainDIP3BeforeActivationSetup */
+            {  430, uint256S("0x20c025ce708233d05280099f82f09def00ed5aaf5430316070e0dd8807897c41") },
+            /* TestChainDIP3Setup */
+            {  431, uint256S("0x7ca56efe77e768762cc94644828b5824a19f77568440abd851efa644079bd6eb") },
+            /* TestChainBRRBeforeActivationSetup */
+            {  497, uint256S("0x1333458713817f35d7c03677755fbcbc3e2fe253b90b5ff32b0187653c4df8db") },
+            /* TestChainDIP3V19Setup */
+            { 1000, uint256S("0x0d6e2df0c68ca0385e48a748776ec582b85dec58ce74e9bf679976138fd6e192") },
+        }
+    };
+
+    {
+        LOCK(::cs_main);
+        auto hash = checkpoints.mapCheckpoints.find(num_blocks);
+        assert(
+            hash != checkpoints.mapCheckpoints.end() &&
+            m_node.chainman->ActiveChain().Tip()->GetBlockHash() == hash->second);
+    }
```

Comment:
To answer a similar kind of question but directed over here, the reason we do this but Bitcoin doesn't is because Bitcoin only uses `TestChain100Setup`, the repurposing of that logic to allow a `num_blocks` length chain is a Dash-ism used within our portions of the test suite.

This PR implements backports that introduces _determinism_ in these chains and the assertion check against tip hash (which would correspond to block height `num_blocks`) is always going to be against block 100 for Bitcoin (see class name) but is variable for Dash (see argument name).

I was against removing the check altogether since enforcing determinism and documenting changes to how regtest might behave (by forcing the recomputation of hardcoded hashes) may be valuable. So, it was decided to see every possible `num_blocks` value set and then add a debug line that prints out the block height and associated expected hash, run the relevant test (using `./test_dash -t [test_name]/[subtest]`) then add those values into a map (the existing `CCheckpointData` struct was useful for that) and then run the assertion check on the tip hash at the relevant height _and fail if no such checkpoint entry was available to compare against_.